In [1]:
import pandas as pd
import sys

sys.path.append('..')
from src.features.augmentation import augment_train_pairs

In [2]:
# ============================================================
# LOAD SPLITS
# ============================================================

train_pairs = pd.read_parquet("../data/processed/train_pairs.parquet")
valid_pairs = pd.read_parquet("../data/processed/valid_pairs.parquet")
test_pairs = pd.read_parquet("../data/processed/test_pairs.parquet")

print(f"Train: {len(train_pairs):,}")
print(f"Valid: {len(valid_pairs):,}")
print(f"Test:  {len(test_pairs):,}")

Train: 120,894
Valid: 15,306
Test:  14,812


In [4]:
# ============================================================
# BEFORE AUGMENTATION
# ============================================================

print("=" * 60)
print("TRAIN BEFORE AUGMENTATION")
print("=" * 60)

print(train_pairs["label"].value_counts())
print("\nPair types:")
print(train_pairs["pair_type"].value_counts())

TRAIN BEFORE AUGMENTATION
label
0    85619
1    35275
Name: count, dtype: int64

Pair types:
pair_type
hard_negative_similar_name           50062
easy_negative_different_public_id    35557
positive_same_public_id              35275
Name: count, dtype: int64


In [5]:
# ============================================================
# APPLY AUGMENTATION ONLY TO TRAIN
# ============================================================

train_augmented = augment_train_pairs(
    train_pairs=train_pairs,
    n_positive_aug=30_000,
    n_hard_negative_aug=30_000,
    random_state=42,
)

print("=" * 60)
print("TRAIN AFTER AUGMENTATION")
print("=" * 60)

print(f"Rows before: {len(train_pairs):,}")
print(f"Rows after:  {len(train_augmented):,}")

print("\nLabel distribution:")
print(train_augmented["label"].value_counts())

print("\nPair types:")
print(train_augmented["pair_type"].value_counts())

TRAIN AFTER AUGMENTATION
Rows before: 120,894
Rows after:  180,894

Label distribution:
label
0    115619
1     65275
Name: count, dtype: int64

Pair types:
pair_type
hard_negative_similar_name           50062
easy_negative_different_public_id    35557
positive_same_public_id              35275
augmented_hard_negative              30000
augmented_positive                   30000
Name: count, dtype: int64


In [6]:
# ============================================================
# AUGMENTED POSITIVE EXAMPLES
# ============================================================

display(
    train_augmented[
        train_augmented["pair_type"] == "augmented_positive"
    ][
        [
            "party_name_1",
            "party_name_2",
            "name_latin_1",
            "name_latin_2",
            "label",
            "pair_type",
        ]
    ]
    .head(30)
)

,party_name_1,party_name_2,name_latin_1,name_latin_2,label,pair_type
2,Ёфтаков Голибчон Аскарходжаевич,e ftakov goibchon askarhodzhaevich,e ftakov golibchon askarhodzhaevich,e ftakov goibchon askarhodzhaevich,1,augmented_positive
9,Саидов Рустам Саидумарович,saidov rustam saidumarovich pao,saidov rustam saidumarovich,saidov rustam saidumarovich pao,1,augmented_positive
19,Давлятов Раҷаб Сориевич,davlyatov raҷab sorievich,davlyatov raҷab sorievich,davlyatov raҷab sorievich,1,augmented_positive
49,Шамсиддинов Уктамҷон Холбердиевич,shamsiddinov uktamҷon holberdievich oao,shamsiddinov uktamҷon holberdievich,shamsiddinov uktamҷon holberdievich oao,1,augmented_positive
52,Саймухудинов Ривочиддин Эмомович,oao sai muhudinov rivochiddin emomovich,sai muhudinov rivochiddin emomovich,oao sai muhudinov rivochiddin emomovich,1,augmented_positive
56,Абдуллоев Неъматулло,oao abdulloev nematullo,abdulloev nematullo,oao abdulloev nematullo,1,augmented_positive
64,Никитин Денис Александрович,NIKITIN DENIS ALEKSANDROVICH,nikitin denis aleksandrovich,NIKITIN DENIS ALEKSANDROVICH,1,augmented_positive
77,Худойқулов Нурали,hudoi kulov nurai,hudoi kulov nurali,hudoi kulov nurai,1,augmented_positive
97,Гао Сзен Чжен,szen chzhen,gao szen chzhen,szen chzhen,1,augmented_positive
102,Забиров Н,zabirov n,zabirov n,zabirov n,1,augmented_positive


In [7]:
# ============================================================
# AUGMENTED HARD NEGATIVE EXAMPLES
# ============================================================

display(
    train_augmented[
        train_augmented["pair_type"] == "augmented_hard_negative"
    ][
        [
            "party_name_1",
            "party_name_2",
            "name_latin_1",
            "name_latin_2",
            "label",
            "pair_type",
        ]
    ]
    .head(30)
)

,party_name_1,party_name_2,name_latin_1,name_latin_2,label,pair_type
1,«ԱՄՓ ՀՈԼԴԻՆԳ» ՍՊԸ,amp holding spy trade,amp holding spy,amp holding spy trade,0,augmented_hard_negative
12,«ԱՄՓ ՀՈԼԴԻՆԳՍ» ՓԲԸ,amp holdings pby plus,amp holdings pby,amp holdings pby plus,0,augmented_hard_negative
17,«ԱՄՓ ՀՈԼԴԻՆԳՍ» ՓԲԸ,trade amp holdings pby,amp holdings pby,trade amp holdings pby,0,augmented_hard_negative
25,«ԱՄՓ ՀՈԼԴԻՆԳՍ» ՓԲԸ,amp holdings pby capital,amp holdings pby,amp holdings pby capital,0,augmented_hard_negative
28,«ԱՄՓ ՀՈԼԴԻՆԳՍ» ՓԲԸ,amp holdings pby investment,amp holdings pby,amp holdings pby investment,0,augmented_hard_negative
30,«ԱՄՓ ՀՈԼԴԻՆԳՍ» ՓԲԸ,amp holdings pby trade,amp holdings pby,amp holdings pby trade,0,augmented_hard_negative
35,«ԱՄՓ ՀՈԼԴԻՆԳՍ» ՓԲԸ,capital amp holdings pby,amp holdings pby,capital amp holdings pby,0,augmented_hard_negative
42,«ԱՄՓ ՀՈԼԴԻՆԳ» ՍՊԸ,amp holding spy group,amp holding spy,amp holding spy group,0,augmented_hard_negative
47,«ԱՄՓ ՀՈԼԴԻՆԳ» ՍՊԸ,investment amp holding spy,amp holding spy,investment amp holding spy,0,augmented_hard_negative
50,«ԱՄՓ ՀՈԼԴԻՆԳ» ՍՊԸ,amp holding spy group,amp holding spy,amp holding spy group,0,augmented_hard_negative


In [8]:
# ============================================================
# SAVE AUGMENTED TRAIN
# ============================================================

save_path = "../data/processed/train_pairs_augmented.parquet"

train_augmented.to_parquet(save_path, index=False)

print(f"Saved to: {save_path}")
print(f"Rows: {len(train_augmented):,}")

Saved to: ../data/processed/train_pairs_augmented.parquet
Rows: 180,894


## Вывод по аугментации

Аугментация была применена только к обучающей выборке. Это позволяет не искажать проверочную и тестовую выборки.

После аугментации размер train увеличился со 120 894 до 180 894 пар. Были добавлены искусственные положительные и сложные отрицательные примеры.

При этом часть искусственных примеров может быть спорной: некоторые искажения слишком сильные, а некоторые отрицательные примеры могут теоретически оказаться вариантами одной сущности. Поэтому текущая аугментация рассматривается как первая версия и будет использоваться осторожно.